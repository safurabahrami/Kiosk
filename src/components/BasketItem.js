import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Typography, Button } from '@material-ui/core';
import InputAdornment from '@material-ui/core/InputAdornment';
import TextField from '@material-ui/core/TextField';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import { getBasketItemByProductId, getProductById, getProductInventoryByProductId } from '../redux/selectors';
import * as Actions from '../redux/actionCreators/Actions';
import { validateAdd, validateRemove } from './BasketControlHelper';
import { view_width } from '../style/style';

const styles = theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  margin: {
    margin: theme.spacing.unit,
  },
  withoutLabel: {
    marginTop: theme.spacing.unit * 3,
  },
  textField: {
    flexBasis: 60,
  },
  productTitle: {
    width: '40%',
    textAlign: 'left',
  },
  button: {
    margin: theme.spacing.unit,
    width: view_width(3),
    height: view_width(3),
  },
});


class BasketItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quantity: 1,
      helperText: '',
    };
  }

    handleChange = (value) => {
      this.setState({
        quantity: value,
      });
    };

    onClickAddToBasket = (basketItem, productId, quantity) => {
      const { updateBasket } = this.props;
      updateBasket(basketItem, productId, quantity);
    }

    onClickRemoveFromBasket= (basketItem, productId, quantity) => {
      const { updateBasket } = this.props;
      updateBasket(basketItem, productId, quantity, 'DECR');
    };

    calcRemainNumberInventory = (inventoryQuantity, basketItem) => {
      if (basketItem && inventoryQuantity) {
        return Number.parseInt(inventoryQuantity, 10) - Number.parseInt(basketItem.quantity, 10);
      }
      return inventoryQuantity || 0;
    }

    render() {
      const { productId, classes, basketItem, inventoryQuantity, product } = this.props;
      const { quantity, helperText } = this.state;
      const remainedNumberInventory = this.calcRemainNumberInventory(inventoryQuantity, basketItem);
      return (
        <div className={classes.root} key={productId}>

          <Typography variant="subtitle1" className={classes.productTitle}>
            {product.name}(s)
          </Typography>

          <Button
                    aria-label="Add"
                    variant="fab"
                    className={classes.button}
                    color="secondary"
                    disabled={!validateRemove(quantity, basketItem)}
                    onClick={() => this.onClickRemoveFromBasket(basketItem, productId, quantity)}
                >
            <RemoveIcon />
          </Button>
          <TextField
                    id="with-adornment"
                    placeholder="quantity"
                    className={classNames(classes.margin, classes.textField, { inputType: classes.inputStyle })}
                    InputProps={{
                      startAdornment: <InputAdornment position="start">#</InputAdornment>,
                      className: classes.textFieldFormLabel,
                    }}
                    type="number"
                    value={quantity}
                    helperText={helperText}
                    onChange={(event) => {
                      if (event.target.value >= 0 && (event.target.value <= remainedNumberInventory
                            || (basketItem && event.target.value <= basketItem.quantity))
                      ) {
                        this.setState({ helperText: '' });
                        return this.handleChange(event.target.value);
                      }
                      this.setState({ helperText: `only ${remainedNumberInventory} items left` });
                      const maxItem = Math.max(basketItem.quantity, remainedNumberInventory);
                      return this.handleChange(maxItem);
                    }}
                />
          <Button
                    color="primary"
                    aria-label="Add"
                    variant="fab"
                    className={classes.button}
                    disabled={!validateAdd(quantity, basketItem, inventoryQuantity)}
                    onClick={() => this.onClickAddToBasket(basketItem, productId, quantity)}
                >
            <AddIcon />
          </Button>

        </div>
      );
    }
}

BasketItem.propTypes = {
  basketItem: PropTypes.object.isRequired,
  product: PropTypes.object.isRequired,
  productId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  inventoryQuantity: PropTypes.number.isRequired,
  updateBasket: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired,
};

const mapStateToProps = (state, props) => ({
  basketItem: getBasketItemByProductId(state, props.productId),
  product: getProductById(state, props.productId),
  inventoryQuantity: getProductInventoryByProductId(state, props.productId),
});
const mapDispatchToProps = dispatch => ({
  updateBasket: (basketItem, productId, quantity, decr) => dispatch(Actions.updateBasket(basketItem, productId, quantity, decr)),
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(BasketItem));
