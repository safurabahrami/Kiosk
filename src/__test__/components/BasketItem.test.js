import React from 'react';
import toJson from 'enzyme-to-json';

import BasketItem from "../../components/BasketItem";
import { Button } from '@material-ui/core';

describe("render BasketItem", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<BasketItem.WrappedComponent
        basketItem={
          {
            quantity:7
          }
        }
        product={{}}
        inventoryQuantity={5}
        updateBasket={() => {}}
    />);
  });

  it("render the component", () => {
    expect(wrapper.length).toEqual(1);
  });
  it("match snapshots", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("textField events when all the inventory is purchased ", () => {
  let wrapper;
  beforeEach(() => 
  {
    wrapper = shallow(<BasketItem.WrappedComponent
        basketItem={
          {
            quantity:7
          }
        }
        product={{}}
        inventoryQuantity={7}
        updateBasket={() => {}}
    />).dive();
  });

  it("match snapshots", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("want to remove 5 items - add disable-remove enable", () => {
    const eventQuantity = { target: { value: 5 } };
    wrapper.find('TextField').at(0).simulate('change', eventQuantity);
    expect(wrapper.state().quantity).toEqual(5);
    let addButton = wrapper.find(Button).at(0);
    let removeButton = wrapper.find(Button).at(1);
    expect(addButton.props().disabled).toBe(true);
    expect(removeButton.props().disabled).toBe(false);
  });

  it("want to insert 10 - add disable-change value to purchased quantity-enable remove", () => {
    const eventQuantity = { target: { value: 10} };
    wrapper.find('TextField').at(0).simulate('change', eventQuantity);
    expect(wrapper.state().quantity).toEqual(7);
    let addButton = wrapper.find(Button).at(0);
    let removeButton = wrapper.find(Button).at(1);
    expect(addButton.props().disabled).toBe(true);
    expect(removeButton.props().disabled).toBe(false);
  });
});

describe("textField events When 3 of 7 inventory has been added", () => {
  let wrapper;
  beforeEach(() => 
  {
    wrapper = shallow(<BasketItem.WrappedComponent
        basketItem={
          {
            quantity:3
          }
        }
        product={{}}
        inventoryQuantity={7}
        updateBasket={() => {}}
    />).dive();
  });

  it("change textbox to 4 - add enable remove disable", () => {
    const eventQuantity = { target: { value: 4 } };
    wrapper.find('TextField').at(0).simulate('change', eventQuantity);
    let addButton = wrapper.find(Button).at(0);
    let removeButton = wrapper.find(Button).at(1);
    expect(wrapper.state().quantity).toEqual(4);
    expect(addButton.props().disabled).toBe(false);
    expect(removeButton.props().disabled).toBe(true);
  });

  it("change textbox to 2 -  add and remove enable", () => {
    const eventQuantity = { target: { value: 2} };
    wrapper.find('TextField').at(0).simulate('change', eventQuantity);
    let addButton = wrapper.find(Button).at(0);
    let removeButton = wrapper.find(Button).at(1);
    expect(wrapper.state().quantity).toEqual(2);
    expect(addButton.props().disabled).toBe(false);
    expect(removeButton.props().disabled).toBe(false);
  });
});

describe("textField events When 0 of 7 inventory has been added", () => {
  let wrapper;
  beforeEach(() => 
  {
    wrapper = shallow(<BasketItem.WrappedComponent
        basketItem={
          {
            quantity:0
          }
        }
        product={{}}
        inventoryQuantity={7}
        updateBasket={() => {}}
    />).dive();
  });

  it("not changing the texbox - add enable remove disable", () => {
    expect(wrapper.state().quantity).toEqual(1);
    let addButton = wrapper.find(Button).at(0);
    let removeButton = wrapper.find(Button).at(1);
    expect(addButton.props().disabled).toBe(false);
    expect(removeButton.props().disabled).toBe(true);
  });
});