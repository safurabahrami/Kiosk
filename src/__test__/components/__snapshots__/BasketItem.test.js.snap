// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render BasketItem match snapshots 1`] = `
<BasketItem
  basketItem={
    Object {
      "quantity": 7,
    }
  }
  classes={
    Object {
      "button": "BasketItem-button-6",
      "margin": "BasketItem-margin-2",
      "productTitle": "BasketItem-productTitle-5",
      "root": "BasketItem-root-1",
      "textField": "BasketItem-textField-4",
      "withoutLabel": "BasketItem-withoutLabel-3",
    }
  }
  inventoryQuantity={5}
  product={Object {}}
  updateBasket={[Function]}
/>
`;

exports[`textField events when all the inventory is purchased  match snapshots 1`] = `
<div
  className="BasketItem-root-1"
>
  <WithStyles(Typography)
    className="BasketItem-productTitle-5"
    variant="subtitle1"
  >
    (s)
  </WithStyles(Typography)>
  <WithStyles(Button)
    aria-label="Add"
    className="BasketItem-button-6"
    color="secondary"
    disabled={false}
    onClick={[Function]}
    variant="fab"
  >
    <pure(RemoveIcon) />
  </WithStyles(Button)>
  <TextField
    InputProps={
      Object {
        "className": undefined,
        "startAdornment": <WithStyles(InputAdornment)
          position="start"
        >
          #
        </WithStyles(InputAdornment)>,
      }
    }
    className="BasketItem-margin-2 BasketItem-textField-4"
    helperText=""
    id="with-adornment"
    onChange={[Function]}
    placeholder="item"
    required={false}
    select={false}
    type="number"
    value={1}
    variant="standard"
  />
  <WithStyles(Button)
    aria-label="Add"
    className="BasketItem-button-6"
    color="primary"
    disabled={true}
    onClick={[Function]}
    variant="fab"
  >
    <pure(AddIcon) />
  </WithStyles(Button)>
</div>
`;
